kvbc/include/v4blockchain/detail/blockchain.h

```diff
@@ -119,6 +122,8 @@ class Blockchain {
   std::shared_ptr<concord::storage::rocksdb::NativeClient> native_client_;
   util::ThreadPool thread_pool_{1};
   std::optional<std::future<BlockDigest>> future_digest_;
+  bool need_compaction_{false};
+  std::mutex compaction_mutex_;
```

Comment:
I don't think we need to protect ```need_compaction_```
It's either we deleteblocks i.e. pruning or adding a block, both operations can't happen in parallel.

Reply:
It is to avoid a theoretical race between the compaction async function resetting the flag and another block being created.